### PR TITLE
Data base schemas updating

### PR DESCRIPTION
**Pull request by DedrickEnc**
This pull request updates the database schema to make data more coherent, some additionnal constraints are added to performs this schema updated, here are task performed in this pull request : 
- The tables `group_invoice`, `group_invoice_item`, `employee_invoice` and `employee_invoice_item`, `payment` are removed.
- some bugs are fixed in cashboxes files
- The integration test is updated to respect new constraints defined in the schemas
- The end to end test is updated to respect new constraint defined in the schemas
- The data.sql is also updated to remove duplications

Thank you for contributing!

Before submitting this pull request, please verify that you have:
- [ ] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

### CODE REVIEW

**Comment by DedrickEnc** [comment]
@sfount @jniles

File: server/models/schema.sql [hunk 1]
```diff
@@ -1,3 +1,4 @@
+drop database bhima_test; create database bhima_test; use bhima_test;
```
Comment (jniles) [inline server/models/schema.sql]: This breaks Travis, which infers the database name from the command line.  It also restricts developers from using database names other than `bhima_test`.  Could you remove these lines?

Comment (DedrickEnc) [inline server/models/schema.sql]: Yes of course, it was just for development purpose.

**Comment by DedrickEnc** [comment]
@jniles
 All test are passed now

File: client/test/e2e/price_list/price_list.spec.js [hunk 2]
```diff
@@ -62,24 +67,24 @@ describe('Price Lists', function () {
 
   it('add price_list_items to a price list', function () {
     element(by.id('price_list_' + priceListID1 )).click();
+
     element(by.id('add_item')).click();
     FU.input('ModalCtrl.data.label', item1.label);
     FU.input('ModalCtrl.data.value', item1.value);
     element(by.id('is_percentage')).click();
+    FU.select('ModalCtrl.data.inventory_uuid', 'First Test Inventory Item');
```
Comment (jniles) [inline client/test/e2e/price_list/price_list.spec.js:75]: :dancer:

File: server/controllers/finance/patientInvoice.js [hunk 3]
```diff
@@ -211,11 +211,23 @@ function create(req, res, next) {
...
     invoice.date = new Date(invoice.date);
+  }else{
+    throw new BadRequest(`An invoice is not allowed to miss a date.`);
+  }
+
+  if(items.length < 1){
+    throw new BadRequest(`An invoice is not allowed to miss a items.`);
+  }
+
+  if(!invoice.description){
```
Comment (jniles) [inline server/controllers/finance/patientInvoice.js]: Does the database not enforce these?  I thought it returned a 400 request already...

File: server/controllers/finance/priceList.js [hunk 4]
```diff
@@ -197,6 +198,17 @@ exports.create = function create(req, res, next) {
     `INSERT INTO price_list_item (uuid, inventory_uuid, price_list_uuid,
     label, value, is_percentage) VALUES ?;`;
 
+
+  if(data.items){
+    var hasInventoryUuid = data.items.every(function (pi) {
```
Comment (jniles) [inline server/controllers/finance/priceList.js]: Isn't this functionality enforced by the database?

File: server/models/schema.sql [hunk 5]
```diff
@@ -98,6 +104,7 @@ CREATE TABLE `cash` (
   `description`     TEXT,
   `is_caution`      BOOLEAN NOT NULL DEFAULT 0,
   PRIMARY KEY (`uuid`),
+  UNIQUE KEY `cash_1` (`reference`, `project_id`),
```
Comment (jniles) [inline server/models/schema.sql:107]: These are set by TRIGGERS.  I think they are guaranteed to be unique.  If they fail, I wonder what happens...

File: server/models/schema.sql [hunk 6]
```diff
@@ -1440,13 +1447,14 @@ DROP TABLE IF EXISTS `project`;
 
 CREATE TABLE `project` (
   `id` SMALLINT(5) UNSIGNED NOT NULL AUTO_INCREMENT,
-  `name` TEXT,
-  `abbr` CHAR(3) DEFAULT NULL,
+  `name` VARCHAR(60) NOT NULL,
+  `abbr` CHAR(3) NOT NULL,
```
Comment (jniles) [inline server/models/schema.sql:1451]: :+1:

File: server/models/schema.sql [hunk 7]
```diff
@@ -1727,20 +1749,21 @@ DROP TABLE IF EXISTS `subsidy`;
...
+  UNIQUE KEY `subsidy_1` (`label`),
   KEY `account_id` (`account_id`),
   FOREIGN KEY (`account_id`) REFERENCES `account` (`id`)
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
 
 
 DROP TABLE IF EXISTS `supplier`;
 CREATE TABLE `supplier` (
-  `uuid` BINARY(16) NOT NULL DEFAULT '',
+  `uuid` BINARY(16) NOT NULL,
```
Comment (jniles) [inline server/models/schema.sql:1766]: :+1:

**Comment by jniles** [comment]
@dedrickenc, I have completed my review.

The constraints look good, and all of them seem to be well formed.  Thanks!

I think the database enforces the `400 BadRequest` errors that you manage manually.  Could you see if you can reproduce any issues you had with these?  It would be far better for the database to enforce data integrity.

File: server/models/schema.sql (hunk 5 above)
Comment (DedrickEnc) [inline server/models/schema.sql:107]: You are right, it is ensured by the triggers but it is not a rule so, it is possible for an other extra component as (Symmetric or an other) to insert a duplication. We are loosing nothing by including this rule, we are safe instead.

File: server/controllers/finance/patientInvoice.js (hunk 3 above)
Comment (DedrickEnc) [inline server/controllers/finance/patientInvoice.js]: I have understood the problem and fix it thanks Jon

File: server/models/schema.sql (hunk 5 above)
Comment (jniles) [inline server/models/schema.sql:107]: Good point, didn't think about the symmetric sync.  LGTM.

**Comment by DedrickEnc** [comment]
@jniles
Suggested changes are added

File: server/controllers/finance/priceList.js [hunk 8]
```diff
@@ -16,6 +16,7 @@
 const db    = require('../../lib/db');
 const uuid  = require('node-uuid');
 var NotFound = require('../../lib/errors/NotFound');
+const BadRequest = require('../../lib/errors/BadRequest');
```
Comment (jniles) [inline server/controllers/finance/priceList.js:19]: You no longer need this import.  This comment will not block the pull request.

**Comment by jniles** [comment]
LGTM.

**Comment by jniles** [comment]
LGTM.

**Pull request merged**